x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/PITAwareQueryClient.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.eql.execution.search;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xpack.core.search.action.ClosePointInTimeAction;
+import org.elasticsearch.xpack.core.search.action.ClosePointInTimeRequest;
+import org.elasticsearch.xpack.core.search.action.ClosePointInTimeResponse;
+import org.elasticsearch.xpack.core.search.action.OpenPointInTimeAction;
+import org.elasticsearch.xpack.core.search.action.OpenPointInTimeRequest;
+import org.elasticsearch.xpack.core.search.action.OpenPointInTimeResponse;
+import org.elasticsearch.xpack.eql.session.EqlSession;
+import org.elasticsearch.xpack.ql.index.IndexResolver;
+
+import static org.elasticsearch.xpack.ql.util.ActionListeners.map;
+import static org.elasticsearch.action.ActionListener.wrap;
+
+/**
+ * Extension of basic query, adding awareness Point-in-Time awareness.
+ * Opens a point-in-time, uses it for all queries and closes it when disposed,
+ * freeing consumer from doing any special management for it.
+ */
+public class PITAwareQueryClient extends BasicQueryClient {
+
+    private String pitId;
+    private final TimeValue keepAlive;
+
+    public PITAwareQueryClient(EqlSession eqlSession) {
+        super(eqlSession);
+        this.keepAlive = eqlSession.configuration().requestTimeout();
+    }
+
+    @Override
+    protected void search(SearchRequest search, ActionListener<SearchResponse> listener) {
+        // no pid, ask for one
+        if (pitId == null) {
+            openPIT(wrap(r -> {
+                pitId = r;
+                searchWithPIT(search, listener);
+            }, listener::onFailure));
+        }
+        else {
+            searchWithPIT(search, listener);
+        }
+    }
+
+    private void searchWithPIT(SearchRequest search, ActionListener<SearchResponse> listener) {
+        // don't increase the keep alive
+        search.source().pointInTimeBuilder(new SearchSourceBuilder.PointInTimeBuilder(pitId, null));
+        // get the pid on each request
+        super.search(search, wrap(r -> {
+                pitId = r.pointInTimeId();
+                listener.onResponse(r);
```

Comment:
What do we do on partial failures (shard failures) ? I wonder if these requests should run with `allow_partial_search_results` set to `false` ?

Reply:
We don't support partial failures (it is set to false, inside `RuntimeUtils#prepareRequest`).

Reply:
cool, thanks